### PR TITLE
LPS-190134 Implements ObjectValidationRuleEngine interface to allow Object Validations Client Extensions be used

### PR DESCRIPTION
**Pull request by mateussgn**
Related issues:
https://liferay.atlassian.net/browse/LPS-190134
https://liferay.atlassian.net/browse/LPS-190409
https://liferay.atlassian.net/browse/LPS-190412
https://liferay.atlassian.net/browse/LPS-190518

Instructions for testing:
The sample for tests is attached to this [ticket](https://liferay.atlassian.net/browse/LPS-190134).
**Since this is a scoped validation sample you must follow these steps strictly**

1. Download the attached file to **/bundles/osgi/client-extensions**.
    1.1. If the client-extension folder does not exist, just start the Portal and the directory should be created.
    1.2. Once the file is in the **/bundles/osgi/client-extensions** directory just wait and it will work automatically.
    
2. Go to **/liferay-portal/workspaces/liferay-sample-workspace**
    2.1. Run the command: `gradlew :client-extensions:liferay-sample-etc-spring-boot:bootRun`
    2.2. It will start the client extension sample.
    2.3. Perform this request in your browser: **http://localhost:58081/ready**. If you see a page that says "READY" it worked.
    
3. Create a new Custom Object Definition named "Employee".

4. Add a new Custom Object Field named "name" from Text type.

5. Add a new Object Validation named "Employee Validation" form type " Liferay Sample Etc Spring Boot Spring Boot Object Validation Rule 1".

6. Active the new Object Validation.

7. Publish the new Employee Custom Object Definition.

8. Create a new Object Entry for the new Employee Custom Object Definition.
    8.1. Fill the name field with the string "Invalid Name".
            8.1.1. The Object Entry must not be created.
    8.2. Fill the name field with any other string.
            8.2.1. The Object Entry must be created.

### CODE REVIEW

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:object

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:object

**Comment by pedro-oliveira446** [comment]
Just starting reviewing :)

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/validation/rule/ObjectValidationRuleEngineRegistryImpl.java [hunk 1]
```diff
@@ -29,14 +32,56 @@ public class ObjectValidationRuleEngineRegistryImpl
...
+				objectValidationRuleEngine -> {
+					boolean allowed = true;
+
+					if (objectValidationRuleEngine instanceof CompanyScoped) {
+						CompanyScoped objectActionExecutorCompanyScoped =
+							(CompanyScoped)objectValidationRuleEngine;
+
+						allowed =
+							objectActionExecutorCompanyScoped.isAllowedCompany(
+								companyId);
```
Comment (pedro-oliveira446) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/validation/rule/ObjectValidationRuleEngineRegistryImpl.java:59]: I think there was some obvious bad copy and paste, could you rename ?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/validation/rule/ObjectValidationRuleEngineRegistryImpl.java [hunk 2]
```diff
@@ -29,14 +32,56 @@ public class ObjectValidationRuleEngineRegistryImpl
...
+							ObjectDefinitionScoped) {
+
+						ObjectDefinitionScoped
+							objectActionExecutorObjectDefinitionScoped =
+								(ObjectDefinitionScoped)
+									objectValidationRuleEngine;
+
+						allowed =
+							objectActionExecutorObjectDefinitionScoped.
+								isAllowedObjectDefinition(objectDefinitionName);
```
Comment (pedro-oliveira446) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/validation/rule/ObjectValidationRuleEngineRegistryImpl.java:72]: Same here.

**Review by pedro-oliveira446 — CHANGES_REQUESTED** [review]

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]
Hey @mateussgn see my comments [here](https://github.com/mateussgn/liferay-portal/pull/104#pullrequestreview-1559814931)


Thanks!

**Comment by mateussgn** [comment]
ci:test:object

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:object

**Review by MarinhoFeliphe — APPROVED** [review]
LGTM!! :tada:

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:object

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/validation/rule/FunctionObjectValidationRuleEngineImpl.java [hunk 3]
```diff
@@ -156,6 +158,10 @@ private JSONObject _getPayloadJSONObject(Map<String, Object> objectEntry) {
 	private JSONFactory _jsonFactory;
 
 	private String _key;
+
+	@Reference
+	private Language _language;
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/validation/rule/FunctionObjectValidationRuleEngineImpl.java:164]: @mateussgn here we don't need use `_language` here we can just pass the `_name`.

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:object

**Comment by victorhcunha** [comment]
This pr is causing failures that were detected in ci

- [LocalFile.ObjectCustomValidation#CanDefineExpressionIsUrl](https://testray.liferay.com/home/-/testray/case_results/2403716294?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2403714585%26testrayRunId%3D2403714720%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
- [LocalFile.ObjectCustomValidation#CanDefineExpressionConditionalLogic](https://testray.liferay.com/home/-/testray/case_results/2403716306?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2403714585%26testrayRunId%3D2403714720%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
- [LocalFile.ObjectCustomValidation#CanDefineExpressionIsAnEmail](https://testray.liferay.com/home/-/testray/case_results/2403715822?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2403714585%26testrayRunId%3D2403714720%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
- [LocalFile.ObjectCustomValidation#CanDefineExpressionMatch](https://testray.liferay.com/home/-/testray/case_results/2403716386?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2403714585%26testrayRunId%3D2403714720%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
- [LocalFile.ObjectCustomValidation#CanDefineExpressionConcat](https://testray.liferay.com/home/-/testray/case_results/2403716064?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2403714585%26testrayRunId%3D2403714720%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
- [LocalFile.ObjectCustomValidation#CanDefineExpressionRange](https://testray.liferay.com/home/-/testray/case_results/2403716313?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2403714585%26testrayRunId%3D2403714720%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
- [LocalFile.ObjectCustomValidation#CanDefineExpressionFutureDatesAndPastDates](https://testray.liferay.com/home/-/testray/case_results/2403716023?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2403714585%26testrayRunId%3D2403714720%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
- [LocalFile.ObjectCustomValidation#CanAddEntry](https://testray.liferay.com/home/-/testray/case_results/2403716403?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2403714585%26testrayRunId%3D2403714720%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)

The tests detected that the other types of validations no longer worked, the same behavior was confirmed by testing manually

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:object

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:object

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:stop

**Comment by mateussgn** [comment]
[LPS-190409 Client Extensions Validations Rules do not need script](https://github.com/liferay-objects/liferay-portal/pull/2367/commits/9d1f8b3493cd37c0dca69183d3fb248566558e6c) and [LPS-190409 Add loading icon while validation is being done](https://github.com/liferay-objects/liferay-portal/pull/2367/commits/add1f74c00128990c6ba0a717b32eb65ed7aa409) were reviewed here: https://github.com/liferay-objects/liferay-portal/pull/2430

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:object

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:object

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by mateussgn** [comment]
ci:stop

**Comment by mateussgn** [comment]
ci:test:sf

**Comment by mateussgn** [comment]
ci:test:object

**Comment by mateussgn** [comment]
ci:test:relevant

**Comment by victorhcunha** [comment]
ci:test:stable

**Comment by mateussgn** [comment]
ci:close

**Comment by mateussgn** [comment]
Sent directly to Brian: https://github.com/brianchandotcom/liferay-portal/pull/139312

**Pull request closed without merging**